torch_geometric/contrib/explain/pgm_explainer.py

```diff
@@ -0,0 +1,428 @@
+                (softmax_pred_perturb[neighbors, target] + pred_threshold) >=
+                softmax_pred[neighbors, target]).cpu()] = 0
+
+            samples.append(seeds)
+            pred_samples.append(sample_bool)
+
+        samples = np.asarray(samples)
+        pred_samples = np.asarray(pred_samples)
+        combine_samples = (samples * 10 + pred_samples) + 1
+
+        neighbors = np.array(neighbors.detach().cpu())
+        data_pgm = pd.DataFrame(combine_samples)
+        data_pgm = data_pgm.rename(columns={
+            0: "A",
+            1: "B"
+        })  # Trick to use chi_square test on first two data columns
+        index_original_to_subgraph = dict(
+            zip(neighbors, list(data_pgm.columns)))
+        index_subgraph_to_original = dict(
+            zip(list(data_pgm.columns), neighbors))
+        p_values = []
+
+        dependent_neighbors = []
+        dependent_neighbors_p_values = []
+        for node in neighbors:
+            if node == index:
+                # null hypothesis is perturbing a particular
+                # node has no effect on result
+                p = 0
+            else:
+                chi2, p, _ = chi_square(
+                    index_original_to_subgraph[node],
+                    index_original_to_subgraph[index], [], data_pgm,
+                    boolean=False, significance_level=significance_threshold)
+            p_values.append(p)
+            if p < significance_threshold:
+                dependent_neighbors.append(node)
+                dependent_neighbors_p_values.append(p)
+
+        pgm_stats = torch.ones(x.size(0), dtype=torch.float)
+        node_mask = torch.zeros(x.size(), dtype=torch.int)
+
+        pgm_stats[neighbors] = torch.tensor(p_values, dtype=torch.float)
+
+        if max_subgraph_size is None:
+            pgm_nodes = dependent_neighbors
+        else:
+            top_p = np.min((max_subgraph_size, len(neighbors) - 1))
+            ind_top_p = np.argpartition(p_values, top_p)[0:top_p]
+            pgm_nodes = [
+                index_subgraph_to_original[node] for node in ind_top_p
+            ]
+        node_mask[pgm_nodes] = 1
+        return node_mask, pgm_stats
+
+    def forward(
+        self,
+        model: torch.nn.Module,
+        x: torch.Tensor,
+        edge_index: torch.Tensor,
+        target: torch.Tensor,
+        index: Optional[int] = None,  # node index
+        **kwargs,
+    ) -> Explanation:
+
+        if isinstance(index, torch.Tensor):
+            if index.numel() > 1:
+                raise NotImplementedError(
+                    f"'{self.__class__.__name}' only supports a single "
+                    f"`index` for now")
+            index = index.item()
+
+        num_samples = kwargs.pop('num_samples', 100)
+        significance_threshold = kwargs.pop('significance_threshold', 0.05)
+        max_subgraph_size = kwargs.get('max_subgraph_size', None)
+        pred_threshold = kwargs.pop('pred_threshold', 0.1)
```

Comment:
Move all these to `__init__` and add documentation. I was considering moving them, but I am not sure how to document them.